Initial nixified build of this binary

Ocramius opened the pull request:
This approach relies on @Hoverbear's excellent guide at https://github.com/Hoverbear/hoverbear.org/blob/417d50e1050cf201c88e74681074803a86ccd018/content/blog/2021-06-25-a-flake-for-your-crate/index.md

Ref: Hoverbear/hoverbear.org#a9abd75063a7eed7bccf9d7918fe1f8561a20c8a

With this change, we can run:

```sh
nix develop
nix flake check
nix build .
nix run .
```

After this, I'll pursue:

 * cleanup
 * dockerization of this built result (put it in a scratch container)

Ocramius committed:
Initial nixified build of this binary

This approach relies on @Hoverbear's excellent guide
at https://github.com/Hoverbear/hoverbear.org/blob/417d50e1050cf201c88e74681074803a86ccd018/content/blog/2021-06-25-a-flake-for-your-crate/index.md

Ref: Hoverbear/hoverbear.org#a9abd75063a7eed7bccf9d7918fe1f8561a20c8a

With this change, we can run:

```sh
nix develop
nix flake check
nix build .
nix run .
```

After this, I'll pursue:

 * cleanup
 * dockerization of this built result (put it in a scratch container)

Ocramius committed:
Docker build process working

Now possible to build `x86_64-linux` docker images (fat, ~45Mb) that have GCC and some dynamic load libraries in them.

To use it:

```sh
nix build .\#haha
docker load < target
docker image ls
```

When investigating with `dive`, multiple GCC / DLib derivations are baked in the docker image: static compilation should
be possible with `x86_64-unknown-linux-gnu`, but it doesn't seem like it's currently supported.

Ocramius committed:
Enabling more build target architecture, but static compilation needs more work

Ref: https://github.com/nix-community/naersk/blob/78789c30d64dea2396c9da516bbcc8db3a475207/examples/static-musl/flake.nix#L1-L49
Ref: https://github.com/nix-community/naersk/commit/380e14395524e2b91f94461a23499527a096f1d1

Ocramius committed:
Make CI use the excellent build setup provided by @DeterminateSystems

Ref: https://github.com/DeterminateSystems/magic-nix-cache-action/tree/749fc5bbc9fa49d60c2b93f6c4bc867b82e1d295
Ref: https://github.com/DeterminateSystems/nix-installer-action/tree/3ebd1aebb47f95493b62de6eec0cac3cd74e50a9

Ocramius committed:
Simplified CI for releases to use the upstream template by Laminas

Ref: https://github.com/laminas/laminas-servicemanager/blob/b38f776535c348a0cd422a73361c7ba9c034b2fd/.github/workflows/release-on-milestone-closed.yml
Ref: https://github.com/laminas/workflow-automatic-releases/tree/3f595792c5bf8cf33393937c346a4f18db2a564f

Ocramius committed:
Removed redundant comment

Ocramius committed:
Removed `musl` target, since it's unavailable: re-formatted everything as per `nixpkgs-fmt`

Note: had to add `nixpkgs.nixpkgs-fmt` to inputs.

That is **NOT** `nixfmt` ( https://github.com/serokell/nixfmt/tree/b3ac0a82870775972c035b426e7503d391b21339#installation )

Ocramius commented on .github/workflows/ci.yml:
```diff
@@ -8,17 +8,8 @@ jobs:
+      - uses: DeterminateSystems/nix-installer-action@v4
+      - uses: DeterminateSystems/magic-nix-cache-action@v2
+      - run: nix flake check
+      - run: nix build .\#haha
```
Name to be adjusted

Ocramius committed:
Adding default exposed port, environment variables to image configuration

These are just defaults that help when getting started: their values can be freely changed by any caller using the container.

Ref: https://discourse.nixos.org/t/dockertools-buildimage-and-user-writable-tmp/5397/8
Ref: https://github.com/NixOS/nixpkgs/blob/a9fea43703c233740cd6200f1acd6af06e160bd0/pkgs/build-support/docker/examples.nix#L323-L338
Ref: https://github.com/moby/moby/blob/daa4618da826fb1de4fc2478d88196edbba49b2f/image/spec/v1.2.md#image-json-description'
Ref: https://ryantm.github.io/nixpkgs/builders/images/dockertools/

Ocramius committed:
Publish built docker image to `ghcr.io`, without the use of BuildX

`docker/build-push-action` does not plan to implement push-only support, so we'll need to build our own.

Ref: https://github.com/blockfrost/blockfrost-backend-ryo/pull/94/files (example)
Ref: https://github.com/blockfrost/blockfrost-backend-ryo/blob/45c31491255f4582ff5614aba846c49ac4085df2/.github/workflows/Docker.yaml (example)
Ref: https://github.com/docker/build-push-action/issues/538
Ref: https://github.com/docker/build-push-action/tree/9311bf5263ae5b36f3ec67aff768790c6e2344ad

Note that we no longer push SemVer-friendly images here: just precise tags.

Ocramius commented on flake.nix:
```diff
@@ -0,0 +1,86 @@
+# https://github.com/Hoverbear/hoverbear.org/blob/417d50e1050cf201c88e74681074803a86ccd018/content/blog/2021-06-25-a-flake-for-your-crate/index.md#flakenix
+{
+  description = "My cute Rust crate!";
```
This obviously requires `README.md` and `Makefile` to be adjusted too

Ocramius replied:
text to be adjusted here too

Ocramius commented on flake.nix:
```diff
@@ -0,0 +1,86 @@
+            overlays = [ self.overlay ];
+          };
+        in
+        {
+          "${cargoToml.package.name}" = pkgs."${cargoToml.package.name}";
+          haha = pkgs.dockerTools.buildImage {
+            name = "oci-srm-server-mock-rust";
+            config = {
+              Cmd =
+                [ "${pkgs.${cargoToml.package.name}}/bin/oci-srm-server-mock" ];
+              Env = [
+                "OCI_SRM_SERVER_MOCK_PORT=80"
+                "OCI_SRM_SERVER_MOCK_BASE_URL=http://oci-srm-server-mock/"
+                "PUNCHOUT_SERVER_LOGIN_URI=http://punchout-server/punch-in?foo=bar&pass=example-supersecret"
+                "PUNCHOUT_SERVER_CONFIRMATION_URI=http://punchout-server/cxml-order-request-endpoint"
+              ];
+              ExposedPorts = { "80/tcp" = { }; };
+            };
+          };
+        });
```
`Dockerfile` can be removed too

Ocramius committed:
Marking `Dockerfile` as unused

Ocramius committed:
Build results should also not be committed

Ocramius committed:
Renaming references to "haha" to be "docker-image"

Ocramius committed:
Adjusted `Makefile` to operate with `nix`

Also, loading + extracting image ID in one line.

Ref: https://github.com/docker/cli/issues/2212#issuecomment-953224385

Ocramius commented on Dockerfile:
```diff
@@ -1,3 +1,4 @@
+# @TODO this file is to be removed: it only exists for reference, until we can get static compilation to work in Nix too
```
To be fixed before merging, IMO, but my nix-fu is too weak today, and I need to figure out how to get `-musl` builds running :)

Ocramius replied:
Ref: https://old.reddit.com/r/NixOS/comments/f0yi3b/how_to_build_a_simple_static_rust_binary_using/

Ocramius committed:
In `Makefile`, variables don't work: inlining an expression to avoid the usage of variables

This is because each line is executed in its own shell.

Ref: https://stackoverflow.com/a/61796076

Ocramius commented:
@ciaranmcnulty just an example, FYI: I removed `docker` from the build-side, and only use it for `load`, `push` and `tag` here, with fully locked dependencies (down to the tiniest detail).

Hope this is useful on your end too :)

Ocramius committed:
Switching to `buildLayeredImage` to have better docker layers and caching

Ref: https://grahamc.com/blog/nix-and-layered-docker-images/
Ref: https://github.com/grahamc/grahamc.github.com/blob/42f6ae9d19d286208249026f07d2dcb7023063e1/_posts/2018-10-01-nix-and-layered-docker-images.md
Ref: https://github.com/grahamc/grahamc.github.com/commit/e1821c7873eaaae2cfd86e0cabd49ccef05764f2

Hoverbear commented on flake.nix:
```diff
@@ -0,0 +1,86 @@
+          '';
+          "${cargoToml.package.name}" = pkgs."${cargoToml.package.name}";
+        });
+      devShell = forAllSystems (system:
```
You may get a warning on more recent nixes, I believe convention has drifted to `devShells.default` (eg https://github.com/DeterminateSystems/nix-installer/blob/8395b40223f56a12f97f29750b6fc8feea80ecb1/flake.nix#L123-L159)

Ocramius replied:
Thanks for poking me about this: agreed, I'm indeed simplifying a lot of the stuff until I get only the bits I understand out of it. I'm indeed getting a warning, which I'll chase later :+1: 

The final result I'll get may not be abstracted, or very reusable, but my current aim is just getting the docker image out of this, for now :D

Ocramius replied:
8af03669e66cbebcd631c835ddca6a4cbc82b846

Everything is a bit more rigid, and probably also wrong, but works for now :D

Ocramius committed:
Correcting location of binary in `Makefile`

The binary is now in a derivation referenced in the default `CMD`, so we don't care about it anymore

Ocramius committed:
Disabled default `actix-web` features to get rid of any `-std` liberaries

Notably, `cargo-why` told that `actix-web -> actix-http -> zstd -> zstd-safe -> zstd-sys` is a
chain that leads to a `-sys` dependency that cannot be statically compiled without `ldd`.

This change fixes that, making static compilation very much possible.

Ocramius committed:
Moved to static compilation, hugely simplifying the flake setup

Ref: https://old.reddit.com/r/NixOS/comments/f0yi3b/how_to_build_a_simple_static_rust_binary_using/
Ref: https://github.com/nix-community/naersk/blob/c6a45e4277fa58abd524681466d3450f896dc094/examples/static-musl/flake.nix

In general, I don't fully understand the whole `default.nix` and `flake.nix` syntax here, but
this change:

 * replaces `naersk` (declares `naersk'`) and creates a cargo build environment for `musl`
 * uses the new build environment for producing a statically linked binary.

At this stage, I'm unsure if `nix build` and `nix flake` commands are sharing the same build environment,
and I may need to get rid of any `default.nix` details later on.

Ocramius committed:
Removed unused `LIBCLANG_PATH`

Ocramius committed:
Removed `default.nix`: no longer relevant, and `nix` commands no longer work outside flake context, in this project

Effectively, this shows that `default.nix` wasn't needed in first place :D

Next time, let's just use the flake instead.

Ocramius committed:
Goodbye `Dockerfile`: you are no longer needed

Ocramius commented:
```
warning: The check omitted these incompatible systems: aarch64-darwin, aarch64-linux, x86_64-darwin
Use '--all-systems' to check all.
```

This is fine

:shipit: 

Also, @Hoverbear, amazing job with the magic cache! :cake:
